### PR TITLE
Refactor FXIOS-9692 - Firefox iOS can'be be compiled using Xcode 16

### PR DESCRIPTION
**Pull request by PARAIPAN9**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9692)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/21311)

## :bulb: Description

- This PR fixes and bypasses the errors that are introduced with `Xcode 16 `and `Swift 6` in order that the project could build successfully.
NOTE**: On `Xcode 16`, `macOS Sequoia` even though the project builds successfully I encounter the following error when I start scrolling through the history: 
`Thread 1: "Fatal: supplied item identifiers are not unique. Duplicate identifiers: {(\n    Storage.Site,\n    Storage.Site,\n    Storage.Site\n)}"`. 
## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/TabManagement/Tab.swift
Comment (PARAIPAN9) [inline firefox-ios/Client/TabManagement/Tab.swift:1119]: I believe that the `evaluateJavaScript` overriden methods should be deleted since they don't do anything apart from calling `super`.

File: firefox-ios/Client/Frontend/Autofill/Address/Edit/UpdatableAddressFields+Decodable.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/Frontend/Autofill/Address/Edit/UpdatableAddressFields+Decodable.swift:8]: Was there a domain/naming conflict here that required adding the `Swift.`?

File: firefox-ios/Client/Frontend/Browser/Tabs/Middleware/TabManagerMiddleware.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/Frontend/Browser/Tabs/Middleware/TabManagerMiddleware.swift:451]: At a quick glance this seems like a significant change (going from MT only to a detached `Task`). Was this added solely to clear up compiler warnings? Or was there a separate issue involved?

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift:156]: This seems fairly significant, am I misreading or is this now potentially updating certificates off the MT? What was the behavior previously?

**Review by mattreaganmozilla — COMMENTED** [review]
Thanks for digging into this @PARAIPAN9. I can't recall if we typically perform these Xcode/Swift compatibility updates in conjunction with updates to our CI? Does automation need to be looped in? (cc @clarmso )

There are a couple changes here which I feel like might be good to get extra eyes on. 👀

File: firefox-ios/Client/Frontend/Autofill/Address/Edit/UpdatableAddressFields+Decodable.swift
Comment (PARAIPAN9) [reply firefox-ios/Client/Frontend/Autofill/Address/Edit/UpdatableAddressFields+Decodable.swift:8]: In `Xcode 16` - `Swift 6` we get an error if we conform to protocols in extensions. For example, in this case we get: `Extension declares a conformance of imported type 'UpdatableAddressFields' to imported protocol 'Decodable'; this will not behave correctly if the owners of 'MozillaAppServices' introduce this conformance in the future`. The compiler suggests to fix this by adding `@retroactive Decodable`, but `@retroactive` is available only in Swift 6, so we would not be able to compile our project for lower Swift versions. By adding `Swift.Decodable` we make both happy.

File: firefox-ios/Client/Frontend/Browser/Tabs/Middleware/TabManagerMiddleware.swift
Comment (PARAIPAN9) [reply firefox-ios/Client/Frontend/Browser/Tabs/Middleware/TabManagerMiddleware.swift:451]: The `undoCloseInactiveTabs()` func still runs on MT. If you got to the definition you can see that it has `@MainActor` attribute set.

File: firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift
Comment (PARAIPAN9) [reply firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionViewController.swift:156]: Yes, you're right the UI is updated potentially off the MT as was already implemented. I'll update it.

File: firefox-ios/Client/Frontend/Autofill/Address/Edit/UpdatableAddressFields+Decodable.swift
Comment (mattreaganmozilla) [reply firefox-ios/Client/Frontend/Autofill/Address/Edit/UpdatableAddressFields+Decodable.swift:8]: Thank you, this was news to me

**Comment by clarmso** [comment]
@PARAIPAN9 Could you please take a look at `bootstrap.sh` and `checkout.sh` under Xcode 16? I tried to bump the Bitrise stack to `osx-xcode-16.0.x`, but the build failed at `./bootstrap.sh` and the corresponding step for Focus. The error occurs at `swift run` from the scripts. Here is the Bitrise pipeline on my attempt to use `osx-xcode-16.0.x` stack.
https://app.bitrise.io/app/6c06d3a40422d10f/pipelines/e8ea3c51-70af-4078-88ae-6185541f465c

Here are the software packages available from the `osx-xcode-16.0.x` and `osx-xcode-15.4.x` stacks. Let me revert back to 15.4 stack for now.
https://stacks.bitrise.io/stack_reports/osx-xcode-16.0.x/
https://stacks.bitrise.io/stack_reports/osx-xcode-15.4.x/

File: firefox-ios/Client/Extensions/UIView+ThemeUUIDIdentifiable.swift
Comment (FilippoZazzeroni) [inline firefox-ios/Client/Extensions/UIView+ThemeUUIDIdentifiable.swift:9]: Just wondering, it is now mandatory from Xcode 16 to prefix the class with the package ?

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift
Comment (FilippoZazzeroni) [inline firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift:80]: Do we need here `withCheckedContinuation` ? the code seems synchronous.

**Review by FilippoZazzeroni — COMMENTED** [review]
Nit just the withCheckedContinuation i think it could be removed, but i may missing something

File: firefox-ios/Client/Extensions/UIView+ThemeUUIDIdentifiable.swift
Comment (PARAIPAN9) [reply firefox-ios/Client/Extensions/UIView+ThemeUUIDIdentifiable.swift:9]: The short response is Yes for extension conformances that are part of a different module. For more details you can check the conversation above that I had with Matt Reagan. There I left more details about the error, etc.

Comment (FilippoZazzeroni) [reply firefox-ios/Client/Extensions/UIView+ThemeUUIDIdentifiable.swift:9]: Ok perfect thank you @PARAIPAN9

File: firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift
Comment (PARAIPAN9) [reply firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift:80]: The `URLSession` delegate method `urlSession(_:didReceive:completionHandler:)` is inherently asynchronous. It is designed to handle authentication challenges that occur during a URL session task. The method expects you to perform some asynchronous work (like validating server trust) and then call the provided completion handler when done. 

What's changed in Swift 6 is that the `completionHandler` needs to be now `@Sendable`. This is the new function signature: `func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping @Sendable (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)`.

To make things simpler I chose the concurrency approach by conforming to: `func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge) async -> (URLSession.AuthChallengeDisposition, URLCredential?)` that uses the async/await concurrency model and used `withCheckedContinuation` as a bridge between the modern async/await concurrency model and traditional completion handler-based asynchronous APIs.

Comment (FilippoZazzeroni) [reply firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift:80]: True i think that approach is great. But i'm not understanding the `withCheckedContinuation` because i see in the closure of the continuation there is no method that pass a completion handler. Sorry to ask again, but i may miss something under the hood, like some Thread implication.

I can compile removing the continuation.
```swift
func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge
    ) async -> (
        URLSession.AuthChallengeDisposition,
        URLCredential?
    ) {
        if let serverTrust = challenge.protectionSpace.serverTrust {
            let certificatesHandler = CertificatesHandler(serverTrust: serverTrust)
            self.completion(certificatesHandler.handleCertificates())
            return (.useCredential, URLCredential(trust: serverTrust))
        } else {
            self.completion(nil)
            return (.cancelAuthenticationChallenge, nil)
        }
    }
```

Comment (PARAIPAN9) [reply firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift:80]: Yes, at a closer look, we don't have any asynchronous operations and indeed we don't need `withCheckedContinuation`. Thanks for the detailed approach, I'll update the code.

Comment (FilippoZazzeroni) [reply firefox-ios/Client/Frontend/TrackingProtection/CertificatesHandler.swift]: Thanks @PARAIPAN9 🙏

**Review by FilippoZazzeroni — APPROVED** [review]
Great work, approving from my side. 🎆

**Comment by PARAIPAN9** [comment]
@clarmso, I Ran the build locally and it was configured successfully. Maybe there is a Bitrise problem, is anyone that can help us with this problem?

**Comment by PARAIPAN9** [comment]
I see that fails with  `Killed: 9` error. Typically indicates that the process was terminated by the operating system, often due to excessive memory usage or other resource constraints. This can happen if the build process exceeds the allocated resources.

**Comment by clarmso** [comment]
> I see that fails with `Killed: 9` error. Typically indicates that the process was terminated by the operating system, often due to excessive memory usage or other resource constraints. This can happen if the build process exceeds the allocated resources.

I tried to compile Focus on Github Actions using Xcode 16 but encountered the same `Killed: 9` error. 😞 
https://github.com/mozilla-mobile/firefox-ios/actions/runs/11145822589/job/30976307144

Here is the spec of the Github Actions runner: https://github.com/actions/runner-images/blob/main/images/macos/macos-14-Readme.md. Let me see why `BrowserKit` is not happy with Xcode 16 for both Firefox and Focus.

**Comment by nbhasin2** [comment]
Hi, we should merge this in as if its working for XCode 15.4 which is what we have on BR at the moment.

Locally it works for XCode 15.4 + XCode 16 as well so the automation for BR can be handled in another PR

cc @clarmso @isabelrios @PARAIPAN9 any objections here?

**Review by nbhasin2 — APPROVED** [review]

**Comment by nbhasin2** [comment]
Note: @OrlaM rolled back sentry so this stays on hold until we bring it to latest version 

Hold please !!!

**Comment by clarmso** [comment]
> any objections here?

No. I agree that the Bitrise and Github Actions work on Xcode 16 can come later.

**Review by dragosb01 — APPROVED** [review]

**Pull request merged**
